exercise/cache/points.py

```diff
@@ -133,21 +210,103 @@ def r_augment(children: List[Dict[str, Any]]) -> None:
         })
 
-        if user.is_authenticated:
-            # Augment submission data.
-            submissions = (
-                user.userprofile.submissions.exclude_errors()
-                .filter(exercise__course_module__course_instance=instance)
-                .prefetch_related('exercise', 'notifications')
-                .only('id', 'exercise', 'submission_time', 'status', 'grade', 'force_exercise_points')
-            )
-            for submission in submissions:
+        if is_authenticated:
+            # Augment deviation data.
+            for deviation in deadline_deviations:
                 try:
-                    tree = self._by_idx(modules, exercise_index[submission.exercise.id])
+                    tree = self._by_idx(modules, exercise_index[deviation.exercise.id])
+                except KeyError:
+                    self.dirty = True
+                    continue
+                entry = tree[-1]
+                entry['personal_deadline'] = (
+                    entry['closing_time'] + datetime.timedelta(minutes=deviation.extra_minutes)
+                )
+                entry['personal_deadline_has_penalty'] = not deviation.without_late_penalty
+
+            for deviation in submission_deviations:
+                try:
+                    tree = self._by_idx(modules, exercise_index[deviation.exercise.id])
                 except KeyError:
                     self.dirty = True
                     continue
                 entry = tree[-1]
+                entry['personal_max_submissions'] = (
+                    entry['max_submissions'] + deviation.extra_submissions
+                )
+
+            # Initialize variables for the submission loop.
+            exercise: BaseExercise = None
+            entry: Dict[str, Any] = None
+            is_better_than: IsBetterThan = None
+            final_submission: Optional[Submission] = None
+            last_submission: Optional[Submission] = None
+
+            def check_reveal_rule() -> None:
+                """
+                Evaluate the reveal rule of the current exercise and ensure
+                that feedback is hidden appropriately.
+                """
+                rule = exercise.active_submission_feedback_reveal_rule
+                state = ExerciseRevealState(entry)
+                is_revealed = rule.is_revealed(state)
+                reveal_time = rule.get_reveal_time(state)
+
+                entry.update({
+                    'best_submission': entry['best_submission'] if is_revealed else last_submission.id,
+                    'points': entry['points'] if is_revealed else 0,
+                    'formatted_points': format_points(entry['points'], is_revealed, False),
+                    'passed': entry['passed'] if is_revealed else False,
+                    'feedback_revealed': is_revealed,
+                    'feedback_reveal_time': reveal_time,
+                })
+
+                for submission in entry['submissions']:
```

Comment:
Looking at this loop, I started to think that this data could have been inserted correctly originally (the line 314 `entry['submissions'].append({`). The reveal rule can be defined with the exercise and the exercise is known in the beginning of the submissions loop on line 286. The reveal rule could be used to hide the points when submissions are inserted to the `entry['submissions']` list.

Reply:
No, the reveal rule must be evaluated after all submissions have been iterated. This is because the `COMPLETION` reveal trigger depends on the submission count and the final points of the exercise. If that trigger option didn't exist, we could do what you said and evaluate the reveal rule right when a new exercise is encountered.

Reply:
Right you are! Then you may fix the commit history and merge we shall!